### PR TITLE
Add `useInterfaces` flag to optionally use interface decoders

### LINKED ISSUE #508: Interface decoders break Amino messages with nested Any protobufs
Right now, the only way to utilize interface decoders is the global `options.interfaces.enabled` flag. When this is used, decoding a protobuf recursively decodes all nested `Any` interfaces inside. This is great UX for viewing the raw data of a message, but breaks Amino signing. Here's why...

This is how Amino signing works right now:
image
- `392`: protobuf messages transformed into Amino messages (`toAmino`)
- `394`: Amino messages sent to wallet/signer for signing
- `398`: signed Amino messages decoded back into protobuf messages (`fromAmino`)
- `402`: TX body protobuf encoded with messages

The `toAmino` function uses its nested types' `_ToAmino` interface encoding functions when interfaces are enabled. The `_ToAmino` function uses `toAmino(decode(nestedAny.value))` on the nested type to get the correct Amino data for it in a switch block. The `decode` function then uses `_InterfaceDecoder`s when `Any`s with interfaces exist.

This path shows that encoding to Amino leads to decoding nested `Any` types. When interfaces are *disabled*, `Any` interfaces are decoded with `Any.decode`, which is the correct type that is expected by the chain. When interfaces are *enabled*, the default behavior becomes decoding with the nested type's decoding function.

I think the default behavior should always be to *not* use the interface, so that messages are decoded into the format expected by the chain. Then there can be an additional flag on the `decode` and `toAmino` functions called `useInterfaces`, which defaults to `false`, and setting it to `true` can then recursively decode `Any`s with interfaces. The key is that this flag is passed from `toAmino` through the `_ToAmino` interface encoder to both the `toAmino` and `decode` functions of nested types so that any recursive `decode` functions can decide to simply use `Any.decode` instead of one the `_InterfaceDecoder` functions.

Simply disabling the interface feature globally is insufficient because there are different times we want to richly decode a protobuf and other times we want to encode protobufs for Amino signing.

### PR DESCRIPTION
**Pull request by NoahSaso**
Resolves #508

This adds a `useInterfaces` flag to the `decode` and `toAmino` functions that allows selective control over whether or not interface decoders are used. The `interfaces.useByDefault` plugin option, which defaults to true, determines the default value for the `useInterfaces` flag.

This is a necessary feature because there are reasons for using and not using interface decoders in different situations. A frontend should not have to choose between always using interface decoders and not having them generated at all.

### CODE REVIEW

File: packages/types/src/telescope.ts [hunk 1]
```diff
@@ -21,6 +21,8 @@ interface TelescopeOpts {
 
     interfaces?: {
         enabled?: boolean;
+        useByDefault?: boolean,
```
Comment (pyramation) [inline packages/types/src/telescope.ts:24]: use what by default? is that enabled by default?

File: packages/ast/src/encoding/proto/implements/decoder.ts [hunk 2]
```diff
@@ -61,7 +61,11 @@ export const createInterfaceDecoderHelper = (
       t.returnStatement(
         t.callExpression(
           t.memberExpression(t.identifier(returnType), t.identifier('decode')),
-          [t.memberExpression(t.identifier('data'), t.identifier('value'))]
+          [
+            t.memberExpression(t.identifier('data'), t.identifier('value')),
+            t.identifier('undefined'),
+            t.booleanLiteral(true)
```
Comment (pyramation) [inline packages/ast/src/encoding/proto/implements/decoder.ts:67]: looks like this is just always true

**Comment by pyramation** [comment]
Great to see you got into context and recursively passing all that down. However, I'm trying to understand why touch so many methods when we can just disable `interfaces.enabled`? Or am I missing something?

Is there a way where we don't need to add a param to every `decode()`? What does `interface.enabled = false` not do that you need?

**Comment by NoahSaso** [comment]
> Great to see you got into context and recursively passing all that down. However, I'm trying to understand why touch so many methods when we can just disable `interfaces.enabled`? Or am I missing something?
> 
> Is there a way where we don't need to add a param to every `decode()`? What does `interface.enabled = false` not do that you need?

see this PR for how the useInterfaces flag gets inserted into the generated types: https://github.com/DA0-DA0/dao-dao-ui/pull/1442

the point is that sometimes we want to use interfaces and other times we don't. disabling interfaces means we never get to use interfaces, though this will encode all amino messages correctly. whereas enabling interfaces means we always have to use interfaces, which means we get automatic decoding of interfaces (which is great for parsing and displaying raw message data) but signing amino messages with nested types breaks. so we want to generate interface decoders but only use them sometimes and not other times, hence the runtime flag.

**Comment by pyramation** [comment]
using the `packages` option to enable and disable certain packages' flags — that may be a solution that can already work.

**Comment by NoahSaso** [comment]
> using the `packages` option to enable and disable certain modules or files with options/flags — that may be a solution that can already work.

hmm, my point still applies even within the same package.

for example: when executing authz's MsgExec from a wallet using amino signing, i need interfaces disabled to properly encode the nested authorization type; however when decoding authz's MsgExec used in a DAO proposal, i need interfaces enabled to automatically detect and decode the nested authorization type used.

even within the same package, there are cases when i need to use interfaces and there are other cases when i don't. thus `interfaces.enabled` is insufficient. i agree that it's a decent amount of code, but i think one flag added to only the necessary functions is the minimum amount of code needed to support this. i'm not sure how you can get around adding a runtime flag if the specific thing needed is the ability to decide whether or not to use interfaces during runtime, but i'm super open to other solutions you come up with! would love something lighter too.

**Comment by pyramation** [comment]
fixing conflicts and made new PR here https://github.com/cosmology-tech/telescope/pull/515

**Comment by pyramation** [comment]
- [ ] fix conflicts
- [ ] make it so that `outputv1` in fixtures doesn't change, by making the flag default to `false`

**Comment by Zetazzz** [comment]
https://github.com/cosmology-tech/telescope/pull/516

Merged in the PR 516, so close this one

**Pull request closed without merging**